File: src/director_ai/core/routing/scorer.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# Director-Class AI — PromptRiskScorer

"""Lightweight prompt-risk scorer.

Combines three cheap signals:

* **Length / structure** — very long prompts, prompts with deep
  nesting or unusual punctuation density, and prompts that contain
  system-style markers ("IGNORE PREVIOUS", "SYSTEM:", …) score
  higher. Closed-form, no dependencies.
* **Sanitiser risk** — the existing
  :class:`InputSanitizer` returns a 0-1 risk score; we pass it
  through when the caller has a sanitiser instance.
* **Injection risk** — optional,
  :class:`InjectionDetector` returns an adversarial-signal score
  that complements the sanitiser.

The three signals are combined as a weighted maximum rather than a
linear blend: a single red flag must not be washed out by three
neutral scores. The default weights assume the sanitiser is
authoritative (weight 0.5), injection is strong (weight 0.35), and
the heuristic is a baseline nudge (weight 0.15) — operators tune
them per deployment.

The scorer is deliberately **synchronous and fast**. The heuristic
runs in ``O(len(prompt))`` with a fixed pattern list; even on a
10 000-character prompt it finishes well under 1 ms, so the
gateway can score every request without paying the cost of the
full scoring pipeline.
"""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass
from typing import Any

logger = logging.getLogger("DirectorAI.Routing.Scorer")

# Cheap-to-match markers of system-prompt exfiltration attempts.
# The list is intentionally narrow — it is not a replacement for
# :class:`InputSanitizer`, only a signal for the length heuristic.
_SYSTEM_STYLE_MARKERS: tuple[re.Pattern[str], ...] = (
    re.compile(r"\bignore (?:all |the |your |previous )", re.IGNORECASE),
    re.compile(r"\bSYSTEM\s*:", re.IGNORECASE),
    re.compile(r"\[\s*system\s*\]", re.IGNORECASE),
    re.compile(r"(?:you are|act as) (?:a|an) (?:admin|root|developer)", re.IGNORECASE),
    re.compile(r"\bdelimiter\s+collision\b", re.IGNORECASE),
)

# Characters whose count feeds the structure risk — unbalanced
# delimiters and bracket soup are common in prompt-injection payloads.
_STRUCTURAL_CHARS = "[]{}<>|`"


@dataclass(frozen=True)
class RiskComponents:
    """Per-signal breakdown — exposed so callers can log which
    channel drove a halt."""

    heuristic: float
    sanitiser: float
    injection: float
    combined: float

    def as_dict(self) -> dict[str, float]:
        return {
            "heuristic": self.heuristic,
            "sanitiser": self.sanitiser,
            "injection": self.injection,
            "combined": self.combined,
        }


class PromptRiskScorer:
    """Produce a ``[0, 1]`` risk score for a prompt.

    Parameters
    ----------
    sanitiser :
        An object exposing ``.score(prompt) -> float`` (the existing
        :class:`InputSanitizer` fits); ``None`` disables the channel.
    injection_detector :
        An object exposing ``.detect(output=None, intent=prompt)``
        returning something with a ``risk`` float attribute. The
        public :class:`InjectionDetector` fits. ``None`` disables
        the channel.
    weights :
        ``(heuristic, sanitiser, injection)``; must sum to 1.0.
    max_safe_length :
        Prompt length at which the length heuristic saturates at 1.0.
    prefer_rust :
        When ``True`` (the default) and ``backfire_kernel`` is
        importable, the system-marker regex walk delegates to
        :class:`backfire_kernel.PiiScanner` so the whole marker set
        is checked in a single pass with a ``RegexSet`` prefilter.
        Set ``False`` to force the pure-Python path for
        benchmarking or in environments without the Rust wheel.
    """

    def __init__(
        self,
        *,
        sanitiser: Any | None = None,
        injection_detector: Any | None = None,
        weights: tuple[float, float, float] = (0.15, 0.5, 0.35),
        max_safe_length: int = 8000,
        prefer_rust: bool = True,
    ) -> None:
        if max_safe_length <= 0:
            raise ValueError(f"max_safe_length must be positive; got {max_safe_length}")
        total = sum(weights)
        if not 0.999 <= total <= 1.001:
            raise ValueError(f"weights must sum to 1.0; got {total}")
        self._sanitiser = sanitiser
        self._injection = injection_detector
        self._weights = weights
        self._max_safe_length = max_safe_length
        self._rust_scanner = _build_rust_marker_scanner() if prefer_rust else None

    def score(self, prompt: str) -> RiskComponents:
        """Return a :class:`RiskComponents` breakdown for ``prompt``.

        Empty or whitespace-only prompts score zero across every
        channel — the scoring pipeline cannot process them anyway
        and a high risk score here would starve real traffic.
        """
        if not prompt or not prompt.strip():
            return RiskComponents(0.0, 0.0, 0.0, 0.0)

        heuristic = self._heuristic(prompt)
        sanitiser = self._sanitiser_signal(prompt)
        injection = self._injection_signal(prompt)
        w_h, w_s, w_i = self._weights
        # A single strong channel must be able to trip the router
        # on its own — weighting only breaks ties between multiple
        # moderate signals. Combined = max of raw channels and the
        # weighted linear combination.
        linear = heuristic * w_h + sanitiser * w_s + injection * w_i
        combined = max(heuristic, sanitiser, injection, linear)
        combined = min(1.0, max(0.0, combined))
        return RiskComponents(heuristic, sanitiser, injection, combined)

    # ------------------------------------------------------------------

    def _heuristic(self, prompt: str) -> float:
        length = len(prompt)
        length_ratio = min(1.0, length / float(self._max_safe_length))
        structural = sum(prompt.count(ch) for ch in _STRUCTURAL_CHARS)
        structural_density = structural / max(length, 1)
        structural_risk = min(1.0, structural_density * 40.0)
        marker_hits = self._count_marker_hits(prompt)
        marker_risk = min(1.0, marker_hits * 0.35)
        risk = max(
            0.4 * length_ratio,
            0.7 * structural_risk,
            marker_risk,
        )
        return min(1.0, risk)

    def _count_marker_hits(self, prompt: str) -> int:
        """Delegate the system-marker regex walk to the Rust scanner
        when available. The scanner reports one match per hit across
        all patterns; we only need a distinct-category count, so
        collapse by category name before returning."""
        if self._rust_scanner is not None:
            matches = self._rust_scanner.scan(prompt)
            if not matches:
                return 0
            return len({m[0] for m in matches})
        return sum(1 for p in _SYSTEM_STYLE_MARKERS if p.search(prompt))

    @property
    def backend(self) -> str:
        """``"rust"`` when the fast-path scanner is active, otherwise
        ``"python"``. Exposed for tests and operator-facing logs."""
        return "rust" if self._rust_scanner is not None else "python"

    def _sanitiser_signal(self, prompt: str) -> float:
        if self._sanitiser is None:
            return 0.0
        try:
            raw = self._sanitiser.score(prompt)
        except Exception:  # pragma: no cover — defensive
            logger.debug("sanitiser.score failed; treating as zero")
            return 0.0
        return _clip01(raw)

    def _injection_signal(self, prompt: str) -> float:
        if self._injection is None:
            return 0.0
        try:
            result = self._injection.detect(output="", intent=prompt)
        except Exception:  # pragma: no cover — defensive
            logger.debug("injection detector failed; treating as zero")
            return 0.0
        risk = getattr(result, "risk", None)
        if risk is None:
            return 0.0
        return _clip01(risk)


def _clip01(value: Any) -> float:
    try:
        as_float = float(value)
    except (TypeError, ValueError):
        return 0.0
    if as_float < 0.0:
        return 0.0
    if as_float > 1.0:
        return 1.0
    return as_float


def _build_rust_marker_scanner() -> Any | None:
    """Return a ``backfire_kernel.PiiScanner`` loaded with the
    system-style marker patterns, or ``None`` when the Rust wheel
    is not importable. The scanner works on any regex batch, not
    just PII — the name reflects its original use case but the
    implementation is a generic ``RegexSet`` walker.
    """
    try:
        from backfire_kernel import PiiScanner as _RustScanner
    except ImportError:
        return None
    # The Python patterns are compiled with ``re.IGNORECASE``. Rust
    # ``regex::RegexSet`` does not carry Python flags, so we prepend
    # the ``(?i)`` inline flag on each source pattern.
    patterns: list[tuple[str, str]] = [
        (f"marker_{i}", "(?i)" + p.pattern) for i, p in enumerate(_SYSTEM_STYLE_MARKERS)
    ]
    try:
        return _RustScanner(patterns)
    except Exception:  # pragma: no cover — defensive
        return None

File: gateway/go/internal/risk/scorer.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial licence available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// Director-Class AI — Go-native prompt risk scorer

// Package risk mirrors the Python PromptRiskScorer heuristic so the
// gateway can refuse obvious attacks before any Python RPC is
// issued. The implementation is deliberately identical to
// ``director_ai/core/routing/scorer.py``'s length / structural /
// marker heuristic so traffic gets the same verdict regardless of
// which tier handles it.
//
// The gateway never sees the sanitiser or injection signals — those
// require the Python model to be loaded — so this Go path is
// heuristic-only. Callers that want the full blend must reach back
// into the Python ``PromptRiskScorer`` over gRPC; this package
// covers the cheap first line of defence.
package risk

import (
	"regexp"
	"strings"
)

// Components mirrors the Python ``RiskComponents`` dataclass. Only
// the heuristic channel is populated here; ``Sanitiser`` and
// ``Injection`` stay at zero.
type Components struct {
	Heuristic float64
	Sanitiser float64
	Injection float64
	Combined  float64
}

// Scorer produces a ``[0, 1]`` risk score from a prompt. Thread
// safe — the regex slices are never mutated after construction.
type Scorer struct {
	markers           []*regexp.Regexp
	structuralChars   string
	maxSafeLength     int
	weights           [3]float64
}

// systemStyleMarkers mirrors ``_SYSTEM_STYLE_MARKERS`` in the Python
// scorer.
var systemStyleMarkers = []string{
	`(?i)\bignore (?:all |the |your |previous )`,
	`(?i)\bSYSTEM\s*:`,
	`(?i)\[\s*system\s*\]`,
	`(?i)(?:you are|act as) (?:a|an) (?:admin|root|developer)`,
	`(?i)\bdelimiter\s+collision\b`,
}

const structuralChars = "[]{}<>|`"

// NewScorer builds a Scorer with the defaults that match the Python
// side: length saturates at 8000 characters, weights sum to 1.
func NewScorer() *Scorer {
	return newScorerWithParams(8000, [3]float64{0.15, 0.5, 0.35})
}

// NewScorerWithMaxLength lets callers tune the length-saturation
// threshold; the weights stay at the Python defaults.
func NewScorerWithMaxLength(maxSafeLength int) (*Scorer, error) {
	if maxSafeLength <= 0 {
		return nil, &invalidParamError{field: "maxSafeLength", value: maxSafeLength}
	}
	return newScorerWithParams(maxSafeLength, [3]float64{0.15, 0.5, 0.35}), nil
}

func newScorerWithParams(maxSafeLength int, weights [3]float64) *Scorer {
	regs := make([]*regexp.Regexp, 0, len(systemStyleMarkers))
	for _, pattern := range systemStyleMarkers {
		// patterns are literals compiled at init; a bad literal
		// would be a bug, not a runtime failure — MustCompile is
		// the right call.
		regs = append(regs, regexp.MustCompile(pattern))
	}
	return &Scorer{
		markers:         regs,
		structuralChars: structuralChars,
		maxSafeLength:   maxSafeLength,
		weights:         weights,
	}
}

// Score returns a ``Components`` record for ``prompt``. Mirrors
// :meth:`PromptRiskScorer.score` — empty or whitespace-only
// prompts return all zeros.
func (s *Scorer) Score(prompt string) Components {
	trimmed := strings.TrimSpace(prompt)
	if trimmed == "" {
		return Components{}
	}
	h := s.heuristic(prompt)
	wH, wS, wI := s.weights[0], s.weights[1], s.weights[2]
	linear := h*wH + 0*wS + 0*wI
	combined := h
	if linear > combined {
		combined = linear
	}
	if combined < 0.0 {
		combined = 0.0
	} else if combined > 1.0 {
		combined = 1.0
	}
	return Components{Heuristic: h, Combined: combined}
}

func (s *Scorer) heuristic(prompt string) float64 {
	length := len(prompt)
	lengthRatio := float64(length) / float64(s.maxSafeLength)
	if lengthRatio > 1.0 {
		lengthRatio = 1.0
	}
	structural := 0
	for _, ch := range prompt {
		if strings.ContainsRune(s.structuralChars, ch) {
			structural++
		}
	}
	density := float64(structural) / float64(maxInt(length, 1))
	structuralRisk := density * 40.0
	if structuralRisk > 1.0 {
		structuralRisk = 1.0
	}
	markerHits := 0
	for _, re := range s.markers {
		if re.MatchString(prompt) {
			markerHits++
		}
	}
	markerRisk := float64(markerHits) * 0.35
	if markerRisk > 1.0 {
		markerRisk = 1.0
	}
	risk := 0.4 * lengthRatio
	if v := 0.7 * structuralRisk; v > risk {
		risk = v
	}
	if markerRisk > risk {
		risk = markerRisk
	}
	if risk > 1.0 {
		risk = 1.0
	}
	return risk
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

type invalidParamError struct {
	field string
	value any
}

func (e *invalidParamError) Error() string {
	return "risk: invalid " + e.field
}
